### DISCUSSION TITLE
Does anyone have any reasonable intutition about the normalization method used for the spectrograms in tts training

### DISCUSSION THREAD

**Discussion by JRMeyer** [coqui-ai/TTS#157] [Archived]
<USER>>>> erogol<br/> <DATE> [March 9, 2020, 3:07pm]<br/> <br/> So there are different TTS libraries out there and I see they all use<br/> different normalization methods for spectrogram normalization in model<br/> training.<br/> <br/> Right now in Mozilla TTS what we do is the following;<br/> <br/> `20 curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz np.log10(np.maximum(min_level, x))`<br/> *slash* [-4,4 *slash* ] assuming minimum db is -100<br/> <br/> [github.com](https://github.com/mozilla/TTS/blob/master/utils/audio.py#L151)<br/> <br/> #### [mozilla/TTS/blob/master/utils/audio.py *slash* #L151](https://github.com/mozilla/TTS/blob/master/utils/audio.py#L151)<br/> <br/> ```<br/> def apply_preemphasis(self, x):<br/> if self.preemphasis == 0:<br/> raise RuntimeError(' [!] Preemphasis is set 0.0.')<br/> return scipy.signal.lfilter([1, -self.preemphasis], [1], x)<br/> <br/> def apply_inv_preemphasis(self, x):<br/> if self.preemphasis == 0:<br/> raise RuntimeError(' [!] Preemphasis is set 0.0.')<br/> return scipy.signal.lfilter([1], [1, -self.preemphasis], x)<br/> <br/> def spectrogram(self, y):<br/> if self.preemphasis != 0:<br/> D = self._stft(self.apply_preemphasis(y))<br/> else:<br/> D = self._stft(y)<br/> S = self._amp_to_db(np.abs(D)) - self.ref_level_db<br/> return self._normalize(S)<br/> <br/> def melspectrogram(self, y):<br/> if self.preemphasis != 0:<br/> D = self._stft(self.apply_preemphasis(y))<br/> ```<br/> <br/> First of all, does anyone see any problem here?<br/> <br/> The only obvious thing above is that preemphasis operation is hard to<br/> recover if you do batch inference. There is no straight forward<br/> implementation of it in CUDA since de-preemphasis operation has a<br/> temporal dependency in itself. Y *slash* <br/> ou can approximate it by using RNN layers but it is slow. Hence, I guess<br/> it makes sense to drop preemphasis. This also makes our model<br/> incompatible with the latest vocoder models.<br/> <br/> I see NVIDIA Tacotron implementation does not use any normalization<br/> except the amp_to_db operation<br/> <br/> [github.com](https://github.com/NVIDIA/tacotron2/blob/master/audio_processing.py#L78)<br/> <br/> #### [NVIDIA/tacotron2/blob/master/audio_processing.py *slash* #L78](https://github.com/NVIDIA/tacotron2/blob/master/audio_processing.py#L78)<br/> <br/> ```<br/> angles = angles.astype(np.float32)<br/> angles = torch.autograd.Variable(torch.from_numpy(angles))<br/> signal = stft_fn.inverse(magnitudes, angles).squeeze(1)<br/> <br/> for i in range(n_iters):<br/> _, angles = stft_fn.transform(signal)<br/> signal = stft_fn.inverse(magnitudes, angles).squeeze(1)<br/> return signal<br/> <br/> def dynamic_range_compression(x, C=1, clip_val=1e-5):<br/> '''<br/> PARAMS<br/> ------<br/> C: compression factor<br/> '''<br/> return torch.log(torch.clamp(x, min=clip_val) curl-run-all.sh discourse.mozilla.org html-to-markdown.sh ordered-posts ordered-posts~ TTS.cdx tts.commands tts-emails.txt TTS.pages tts-telegram.txt TTS.warc.gz C)<br/> <br/> def dynamic_range_decompression(x, C=1):<br/> '''<br/> ```<br/> <br/> ESPNet, on the other hand, uses Standardization with mean and variance.<br/> It is good to compute normalization parameters from the target dataset<br/> (as in image recognition) to make the normalization flexible among<br/> different datasets. However, the downside is that each frequency level<br/> attains the same level of consideration by the model. I don't think it<br/> is the right thing to do since for speech different frequency levels<br/> signify different aspects of the speech. Another downside is that in a<br/> multi-speaker model we need to compute mean-var stats separately per<br/> speaker. Which is a viable but an additional headache.<br/> <br/> Also, I saw that using Standardization enables better vocoder models,<br/> especially with the new GAN based models.<br/> <br/> I also started to experiment with Standardization and saw that the<br/> training seems more stable but the GL based results sound worse.<br/> <br/> So I guess the better option is to use Standardization with a trained<br/> vocoder and our current normalization flow for GL.<br/> <br/> These are all I know and I am kind of confused here. Please let me know<br/> if you have any take in this issue.<br/> <br/> [This is an archived TTS discussion thread from discourse.mozilla.org/t/does-anyone-have-any-reasonable-intutition-about-the-normalization-method-used-for-the-spectrograms-in-tts-training]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> nmstoker<br/> <DATE> [March 15, 2020, 4:32pm]<br/> <br/> This isn't something I know about particularly but I did start reading<br/> around to see what others discussed etc. It might help to visualise the<br/> impact of different approaches on the spectrograms / audio waves before<br/> & after.<br/> <br/> Regarding dropping the pre-emphasis there were mentions in the context<br/> of speech recognition that it possibly wasn't necessary any longer:<br/> <https://www.quora.com/Why-is-pre-emphasis-i-e-passing-the-speech-signal-through-a-first-order-high-pass-filter-required-in-speech-processing-and-how-does-it-work> *slash* <br/> I'd be inclined to test it w/o, and it sounds like others most not be<br/> doing it too, given what you say about other vocoders.<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> dkreutz<br/> <DATE> [March 15, 2020, 8:10pm]<br/> <br/> Yes, Pre-emphasis can be skipped.<br/> <br/> Regarding normalization see<br/> <https://en.wikipedia.org/wiki/Audio_normalization><br/> <br/> From what I understand the peak normalization approach is used here.<br/> Probably a loudness normalization (e.g. RMS) would be more feasible as<br/> the average level 'seen by the algorithm' is more constant?<br/> <br/> [Archived Post]<br/>

**Comment by JRMeyer** [comment]
<USER>>>> erogol<br/> <DATE> [March 17, 2020, 11:28am]<br/> <br/> the problem with preemphasis is practical. Yes, it improves the results<br/> but it is hard to apply in inference time since there is no GPU<br/> implementation.<br/> <br/> My latest experiments showed mean-var normalization works a bit better<br/> than the other methods. Now I am adding it to our repo as another<br/> alternative. You can also still use preemphasis as you like. I don't<br/> drop the support for it.<br/> <br/> [Archived Post]<br/>
